### ISSUE TITLE
[BUG] Opensearch SSL transport error, master not discovered or elected yet

### ISSUE THREAD

**Issue by alborotogarcia** [opensearch-project/helm-charts#54]
**Describe the bug**
Can't reproduce default demo setup on kubernetes.

**To Reproduce**
1. Install helm chart with defaults (optional) from https://github.com/opensearch-project/helm-charts
2. Copy all configuration yaml from /usr/share/opensearch/plugins/opensearch-security/securityconfig to local
3. Paste contents to securityConfig.config.data file templates
```
SSLHandshakeException: Insufficient buffer remaining for AEAD cipher fragment (2). Needs to be more than tag size (16)
[opensearch-cluster-master-0] master not discovered or elected yet
```

**Expected behavior**
Cluster gets GREEN state

**Plugins**
Please list all plugins currently enabled.
```opensearch.yml:
    cluster.name: opensearch-cluster

    # Bind to all interfaces because we don't know what IP address Docker will assign to us.
    network.host: 0.0.0.0

    # # minimum_master_nodes need to be explicitly set when bound on a public IP
    # # set to 1 to allow single node clusters
    discovery.zen.minimum_master_nodes: 1
    plugins:
      security:
        ssl:
          transport:
            pemcert_filepath: esnode.pem
            pemkey_filepath: esnode-key.pem
            pemtrustedcas_filepath: root-ca.pem
            enforce_hostname_verification: false
          http:
            enabled: false
            pemcert_filepath: esnode.pem
            pemkey_filepath: esnode-key.pem
            pemtrustedcas_filepath: root-ca.pem
        allow_unsafe_democertificates: true
        allow_default_init_securityindex: true
        authcz:
          admin_dn:
            - CN=kirk,OU=client,O=client,L=test, C=de
        audit.type: internal_opensearch
        enable_snapshot_restore_privilege: true
        check_snapshot_restore_write_privileges: true
        restapi:
          roles_enabled: ["all_access", "security_rest_api_access"]
        system_indices:
          enabled: true
          indices:
            [
              ".opendistro-alerting-config",
              ".opendistro-alerting-alert*",
              ".opendistro-anomaly-results*",
              ".opendistro-anomaly-detector*",
              ".opendistro-anomaly-checkpoints",
              ".opendistro-anomaly-detection-state",
              ".opendistro-reports-*",
              ".opendistro-notifications-*",
              ".opendistro-notebooks",
              ".opendistro-asynchronous-search-response*",
            ]
```

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

**Comment by peterzhuamazon** [comment]
Never seen this issue before from me, @DandyDeveloper @TheAlgo any idea on this issue from @alborotogarcia ?
Thanks.

**Comment by DandyDeveloper** [comment]
That specific bug can be ignored, its a known thing in Search Guard: https://bugs.openjdk.java.net/browse/JDK-8221218

Shouldn't have any impact on the cluster working. 

> [opensearch-cluster-master-0] master not discovered or elected yet

Is this actually causing problem? You mention the cluster being green. 

If you are just trying to run a single cluster; 

```sh

    # # minimum_master_nodes need to be explicitly set when bound on a public IP
    # # set to 1 to allow single node clusters
    # discovery.zen.minimum_master_nodes: 1

    # Setting network.host to a non-loopback address enables the annoying bootstrap checks. "Single-node" mode disables them again.
    #discovery.type: single-node
```

Uncomment these it'll work. 

If not, we need the full log.

**Comment by alborotogarcia** [comment]
I meant green state as desired, not really reached unfortunately, as securityconfig doesn't get started
I've created a gist with the values.yaml and the full trace of the three nodes [here](https://gist.github.com/alborotogarcia/427f996665b591e62db30a7fc6b4ba47) , please could you take a look @DandyDeveloper ? 
If there's something else I'm missing let me know :)

Thanks for the help @peterzhuamazon @DandyDeveloper !

**Comment by alborotogarcia** [comment]
FWIW @DandyDeveloper @peterzhuamazon , I forgot to mention, internal users and other configurations added works if they're kept in their volumes and I redeploy the helm chart one more time with no securityConfig.config.data.

**Comment by smlx** [comment]
this seems to be the problem
```
opensearch-cluster-master-0 opensearch java.nio.file.FileSystemException: /usr/share/opensearch/data/nodes/0/.opensearch_temp_file: Read-only file system
```

**Comment by alborotogarcia** [comment]
@smlx I see, since kubernetes version 1.9.6 and forth, volumeMounts behavior on secret, configMap, downwardAPI and projected have changed to Read-Only by default as stated here https://github.com/kubernetes/kubernetes/issues/62099 But I don't understand why just leaving as the default chart template it doesn't complain about RO filesystem.. is it another UID that initiates the process? the current fsGroup is set to user 1000 and so it is set on https://github.com/opensearch-project/helm-charts/pull/9
How can this be solved?

**Comment by DandyDeveloper** [comment]
@alborotogarcia 

I just deployed locally with your exact values and its working for me and able to write to that directory. 

```sh
[opensearch@opensearch-cluster-master-0 ~]$ cd data/
[opensearch@opensearch-cluster-master-0 data]$ ls -l
total 20
-rw-rw-r-- 1 opensearch opensearch    5 Sep 24 01:29 batch_metrics_enabled.conf
-rw-rw-r-- 1 opensearch opensearch    5 Sep 24 01:29 logging_enabled.conf
drwxrwxr-x 3 opensearch opensearch 4096 Sep 24 01:29 nodes
-rw-rw-r-- 1 opensearch opensearch    5 Sep 24 01:29 performance_analyzer_enabled.conf
-rw-rw-r-- 1 opensearch opensearch    5 Sep 24 01:29 rca_enabled.conf
[opensearch@opensearch-cluster-master-0 data]$ ls -l nodes/
total 4
drwxrwxr-x 3 opensearch opensearch 4096 Sep 24 01:39 0
[opensearch@opensearch-cluster-master-0 data]$ ls -l nodes/0
total 4
drwxrwxr-x 2 opensearch opensearch 4096 Sep 24 01:29 _state
-rw-rw-r-- 1 opensearch opensearch    0 Sep 24 01:29 node.lock
```

In relation to your specific issue though, I am a bit concerned we're not using `subPath` to define the specific files in the `data` dir:

```
        volumeMounts:
        - mountPath: /usr/share/opensearch/data
          name: opensearch-cluster-master
...
      volumes:
      - configMap:
          defaultMode: 420
          name: opensearch-cluster-master-config
        name: config
```

Because yes, on the surface, I would expect this to report read-only FS because we're mounting a configMap to the data dir.

Maybe we at the least to alleviate concerns add a subPath seeing as only 1 file is written in this configMap anyway;

```sh
apiVersion: v1
data:
  opensearch.yml: |
```

**Comment by alborotogarcia** [comment]
Sorry for the delay @DandyDeveloper, I had some issues with my IdP and had to spent time on it.. I am running a 3 node k3s cluster and yes I am aware that all config files are needed otherwise it will complain.. I run longhorn as storage class.. but IMHO I suspect that If I turn it to subpaths for each file mounts it may be less error prone.. as you said earlier it may affect to the folder that it gets mounted on.. will report back

**Comment by mprimeaux** [comment]
@DandyDeveloper We are running into what I perceive as the same or similar issue with the `1.0.0` charts with a similar config as @alborotogarcia, though we are using Keycloak as our idP. 

Would you mind reviewing the permissions in the `/usr/share/opensearch/plugins/opensearch-security` folder?  It appears the `securityconfig` is owned by `root` and not `opensearch`, which might be the cause.

```shell
-rw-r--r-- 1 opensearch opensearch  452868 Jul  8 22:32 saaj-impl-1.5.2.jar
drwxrwsrwt 3 root       opensearch     260 Sep 26 12:24 securityconfig
-rw-r--r-- 1 opensearch opensearch   41203 Jul  8 22:32 slf4j-api-1.7.25.jar
```

**Comment by mprimeaux** [comment]
It seems my previous assumption is incorrect. Applying an older version of the OpenSearch Helm chart with the same values file works even with the same folder and file permissions as above.

``shell
-rw-r--r-- 1 opensearch opensearch  452868 Jul  8 22:32 saaj-impl-1.5.2.jar
drwxrwsrwt 3 root       opensearch     260 Sep 26 12:44 securityconfig
-rw-r--r-- 1 opensearch opensearch   41203 Jul  8 22:32 slf4j-api-1.7.25.jar
```

```shell
[opensearch@opensearch-cluster-master-0 securityconfig]$ ls -l
total 0
lrwxrwxrwx 1 root opensearch 24 Sep 26 12:44 action_groups.yml -> ..data/action_groups.yml
lrwxrwxrwx 1 root opensearch 16 Sep 26 12:44 audit.yml -> ..data/audit.yml
lrwxrwxrwx 1 root opensearch 17 Sep 26 12:44 config.yml -> ..data/config.yml
lrwxrwxrwx 1 root opensearch 25 Sep 26 12:44 internal_users.yml -> ..data/internal_users.yml
lrwxrwxrwx 1 root opensearch 19 Sep 26 12:44 nodes_dn.yml -> ..data/nodes_dn.yml
lrwxrwxrwx 1 root opensearch 16 Sep 26 12:44 roles.yml -> ..data/roles.yml
lrwxrwxrwx 1 root opensearch 24 Sep 26 12:44 roles_mapping.yml -> ..data/roles_mapping.yml
lrwxrwxrwx 1 root opensearch 18 Sep 26 12:44 tenants.yml -> ..data/tenants.yml
lrwxrwxrwx 1 root opensearch 20 Sep 26 12:44 whitelist.yml -> ..data/whitelist.yml
[opensearch@opensearch-cluster-master-0 securityconfig]$
```

I'll continue debugging.

**Comment by mprimeaux** [comment]
When using the latest version of the OpenSearch chart with the same values file as above, these are the exceptions we receive, which prevent `security_admin` from succeeding:

<details>
  <summary>Error</summary>
  <p>

```shell
opensearch [2021-09-26T12:26:49,688[][DEBUG[][o.o.s.c.ConfigurationRepository[] [opensearch-cluster-master-0[] Try to load config ...
opensearch [2021-09-26T12:26:49,689[][DEBUG[][o.o.s.c.ConfigurationRepository[] [opensearch-cluster-master-0[] security index not exists (yet)
opensearch [2021-09-26T12:26:49,689[][ERROR[][o.o.s.c.ConfigurationLoaderSecurity7[] [opensearch-cluster-master-0[] Exception while retrieving configuration for [INTERNALUSERS, ACTIONGROUPS, CONFIG, ROLES, ROLESMAPPING, TENANTS, NODESDN, WHITELIST, AUDIT[] (index=.opendistro_security)
opensearch org.opensearch.cluster.block.ClusterBlockException: blocked by: [SERVICE_UNAVAILABLE/1/state not recovered / initialized];
opensearch     at org.opensearch.cluster.block.ClusterBlocks.globalBlockedException(ClusterBlocks.java:203) ~[opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.cluster.block.ClusterBlocks.globalBlockedRaiseException(ClusterBlocks.java:189) ~[opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.action.get.TransportMultiGetAction.doExecute(TransportMultiGetAction.java:72) ~[opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.action.get.TransportMultiGetAction.doExecute(TransportMultiGetAction.java:53) ~[opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:192) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.indexmanagement.rollup.actionfilter.FieldCapsFilter.apply(FieldCapsFilter.kt:141) [opensearch-index-management-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:190) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.security.filter.SecurityFilter.apply0(SecurityFilter.java:234) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.security.filter.SecurityFilter.apply(SecurityFilter.java:154) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:190) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.performanceanalyzer.action.PerformanceAnalyzerActionFilter.apply(PerformanceAnalyzerActionFilter.java:99) [opensearch-performance-analyzer-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:190) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.action.support.TransportAction.execute(TransportAction.java:168) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.action.support.TransportAction.execute(TransportAction.java:96) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.client.node.NodeClient.executeLocally(NodeClient.java:99) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.client.node.NodeClient.doExecute(NodeClient.java:88) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.client.support.AbstractClient.execute(AbstractClient.java:428) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.client.support.AbstractClient.multiGet(AbstractClient.java:546) [opensearch-1.0.0.jar:1.0.0[]
opensearch     at org.opensearch.security.configuration.ConfigurationLoaderSecurity7.loadAsync(ConfigurationLoaderSecurity7.java:211) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.security.configuration.ConfigurationLoaderSecurity7.load(ConfigurationLoaderSecurity7.java:102) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.security.configuration.ConfigurationRepository.getConfigurationsFromIndex(ConfigurationRepository.java:375) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.security.configuration.ConfigurationRepository.reloadConfiguration0(ConfigurationRepository.java:321) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.security.configuration.ConfigurationRepository.reloadConfiguration(ConfigurationRepository.java:306) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at org.opensearch.security.configuration.ConfigurationRepository$1.run(ConfigurationRepository.java:166) [opensearch-security-1.0.0.0.jar:1.0.0.0[]
opensearch     at java.lang.Thread.run(Thread.java:832) [?:?]
```

</p>
</details>

If this turns out to be a different issue than the issue that's the topic of this thread then I'll open a separate issue.

**Comment by mprimeaux** [comment]
I believe I found the issue. 

It appears the behavior of the `majorVersion` chart value changed from computing the value `7` to a value of `1`. The workaround (for me, anyway) was to explicitly set the `majorVersion` in the values file to `7`. i.e. ` majorVersion: 7`

If the `majorVersion` attribute remains at its default of `""`, then the stateful set computes the `env:` stanza as:

```yaml
- name: discovery.zen.minimum_master_nodes
  value: "1"
- name: discovery.zen.ping.unicast.hosts
  value: "opensearch-cluster-master-headless"
```

...rather than:

```yaml
- name: cluster.initial_master_nodes
  value: "opensearch-cluster-master-0,opensearch-cluster-master-1,opensearch-cluster-master-2,"
- name: discovery.seed_hosts
  value: "opensearch-cluster-master-headless"
```

When using "discovery", the failures as per [above](https://github.com/opensearch-project/helm-charts/issues/54#issuecomment-927301658) are present and the security indexes are never created thus resulting in a red cluster status.

I am not familiar with zen discovery. All thoughts and support are welcome.

**Comment by alborotogarcia** [comment]
@mprimeaux @DandyDeveloper I followed your suggestions, and here's what it worked for me
```
discovery.zen.minimum_master_nodes: 1
discovery.seed_hosts: "opensearch-cluster-master-headless"
```
and let majorVersion: ""

though I can't still login with my IdP

**Comment by TheAlgo** [comment]
@mprimeaux @alborotogarcia I did not try out the config and installation as I am away from work for some time. But I am thinking out loud. Can this be something related to the core engine and not the chart? Maybe we might need to look at the security repository to understand more because ideally `7` should not fix the issue as OpenSearch starts with `1`.

**Comment by mprimeaux** [comment]
@TheAlgo [Here](https://github.com/opensearch-project/helm-charts/blob/dd0c692a9ca1dc97baa5931b67187bdfc333824a/charts/opensearch/templates/statefulset.yaml#L295) is logic. It *appears* to be, in part, an issue with the chart logic since we [SHOULD](https://tools.ietf.org/html/rfc2119) be using different discovery `env:` values. 

However, I agree with you that something deeper is likely going on and so I will also research the security repository.

Related to the [OpenDistro docs](), it seems they are stale given the discovery attributes are [discovery.zen.ping.unicast.hosts](https://github.com/opensearch-project/OpenSearch/blob/46e0f63539e8ae5bc8f7736c4070dcaa2d0b281e/server/src/main/java/org/opensearch/discovery/SettingsBasedSeedHostsProvider.java#L61) and [discovery.see_hosts](https://github.com/opensearch-project/OpenSearch/blob/46e0f63539e8ae5bc8f7736c4070dcaa2d0b281e/server/src/main/java/org/opensearch/discovery/SettingsBasedSeedHostsProvider.java#L64). 

I did find [this](https://opensearch.org/docs/opensearch/cluster/#step-4-configure-discovery-hosts-for-a-cluster) in the OpenSearch docs.

**Comment by mprimeaux** [comment]
@alborotogarcia Thanks, mate. I will try your suggestion in the above [reply](https://github.com/opensearch-project/helm-charts/issues/54#issuecomment-927336270).

**Comment by TheAlgo** [comment]
> @TheAlgo [Here](https://github.com/opensearch-project/helm-charts/blob/dd0c692a9ca1dc97baa5931b67187bdfc333824a/charts/opensearch/templates/statefulset.yaml#L295) is logic. It _appears_ to be, in part, an issue with the chart logic since we [SHOULD](https://tools.ietf.org/html/rfc2119) be using different discovery `env:` values.
> 
> However, I agree with you that something deeper is likely going on and so I will also research the security repository.
> 
> Related to the OpenDistro docs, it seems they are stale given the discovery attributes are [discovery.zen.ping.unicast.hosts](https://github.com/opensearch-project/OpenSearch/blob/46e0f63539e8ae5bc8f7736c4070dcaa2d0b281e/server/src/main/java/org/opensearch/discovery/SettingsBasedSeedHostsProvider.java#L61) and [discovery.see_hosts](https://github.com/opensearch-project/OpenSearch/blob/46e0f63539e8ae5bc8f7736c4070dcaa2d0b281e/server/src/main/java/org/opensearch/discovery/SettingsBasedSeedHostsProvider.java#L64) as per [this](https://opensearch.org/docs/opensearch/cluster/#step-4-configure-discovery-hosts-for-a-cluster) in the OpenSearch docs.

@mprimeaux We need to change the Helm logic for sure.  As part of #21 we changed it at 1 place and did not change the others which seemed to breaking.

Coming to the `OpenDistro` docs , yes it is stale and we should follow the official OpenSearch docs as much as possible.

**Comment by mprimeaux** [comment]
It appears the following `env:` stanza used in OpenSearch testing at [ZenDiscoveryUnitTests.java](https://github.com/opensearch-project/OpenSearch/blob/0ba0e7cc26060f964fcbf6ee45bae53b3a9941d0/server/src/test/java/org/opensearch/discovery/zen/ZenDiscoveryUnitTests.java#L284) is being deprecated.

See the cluster settings logic [here](https://github.com/opensearch-project/OpenSearch/blob/072ccda932b40ab4f5a3bd248ca91fb8c2aba5b6/buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java#L352).  I believe this might be a point of focus for the chart logic.

**Comment by alborotogarcia** [comment]
@DandyDeveloper Also an ingress api upgrade from networking.k8s.io/v1beta1 to networking.k8s.io/v1 on kubernetes 1.22+

**Comment by mprimeaux** [comment]
@alborotogarcia Coincidentally, I noticed this also, *and*, in addition it seems `IngressClassName` support was removed from the latest OpenSearch charts when migrated from the old repository. The ingress template should add this back as per [Kubernetes 1.18+](https://kubernetes.io/docs/concepts/services-networking/ingress/#deprecated-annotation):

```yaml
  {{- if and .Values.ingress.ingressClassName }}
  ingressClassName: {{ .Values.ingress.ingressClassName | quote }}
  {{- end }}
```

I will create a new issue and related PR today for the `IngressClassName` to be supported. But this is unrelated to this current issue.

**Comment by peterzhuamazon** [comment]
Close this for now as it seems to be resolved by community.
Please feel free to re-open if you still have questions.

Thanks.

**Comment by dss010101** [comment]
Im seeing this error when trying to stand up this infrastructure with the latest images:  https://github.com/opensearch-project/data-prepper/tree/main/examples/log-ingestion

And using this fluentbit.conf:
```

[SERVICE]
  Flush                 1
  Log_Level             debug
  parsers_file          parsers.conf
  trace_error           On

[INPUT]
  name                  tail
  tag                   app_svc_logs
  refresh_interval      5
  path                  /var/log/test.log
  path_key              file_name
  read_from_head        true
  multiline.parser      python

[OUTPUT]
  Name opensearch
  Match *
  host opensearch
  tls On
```
  
  It is not clear to be from this thread what modifications i can make to the docker-compose or other settings to resovle this.  this is the error:

`opensearch             | [2023-05-12T23:08:01,314][WARN ][o.o.h.AbstractHttpServerTransport] [0248e6704e54] caught exception while handling client http traffic, closing connection Netty4HttpChannel{localAddress=/192.168.32.2:9200, remoteAddress=/192.168.32.4:35794}`
